core/src/main/scala/kafka/quota/ClientQuotaCallback.scala

```diff
@@ -0,0 +1,139 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.quota
+
+import kafka.network.RequestChannel.Session
+import org.apache.kafka.common.{Configurable, TopicPartition}
+
+/**
+ * Quota callback interface for clients that enables customization of quota computation.
+ */
+trait ClientQuotaCallback extends Configurable {
+
+  /**
+   * Quota callback invoked to determine the quota limit to be applied for a request.
+   *
+   * @param session The session for which quota is requested
+   * @param clientId The client id associated with the request
+   * @param quotaType Type of quota requested
+   *
+   * @return the quota including the limit and metric tags that indicate which other entities share this quota
+   */
+  def quota(session: Session, clientId: String, quotaType: ClientQuotaType): ClientQuota
```

Comment:
@junrao We need to go through the quota configs at different levels to determine the metric tags. `quotaLimit` also does something similar. So it is cheaper to return the limit with the tags, and since this is used for every request, it may be worth returning limit and tags together. With two calls, we also need to handle changes between the two invocations. Alternatively, we could do something different for default callback handler alone since custom handlers are less likely to have several levels. What do you think?

Reply:
@rajinisivaram : Hmm, my understanding is that quota() only needs to be called when the quota metric doesn't already exist. So, typically, quota() is only called on the first request from a given client and not on every request.

Reply:
@junrao At the moment, we call `quota()` on every request since need to work out which metric tags to use and to do that we need to know which level of configs are defined. But, you are right, after the first request, we already know the limit since the metric exists, so we don't need to call two methods every time. So agree, we can change the first method to just return metric tags. Will update the PR, thanks.

Reply:
@junrao I have changed the first method to `quotaMetricTags` to obtain just the tags and updated the code to use `quotaLimit` to get the limit in all cases. Thanks.